### PR TITLE
Add server implementation of refrigerator alarm cluster

### PR DESCRIPTION
**Pull request by jadhavrohit924**
- Changes

> Server implementation of refrigerator alarm.
> Zap generated

### CODE REVIEW

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple Please review src/app/clusters code. Will rebase this once #26572 get merged.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:35]:
```suggestion
static Status resetHandler(CommandHandler * commandObj, const ConcreteCommandPath & commandPath, const BitMask<AlarmMap> & alarms, const Optional<BitMask<AlarmMap>> & mask)
```
though for the BitMask maybe the ref is not needed since it might end up just word-sized... worth checking.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:56]: This seems like a somewhat complicated way to write:
```suggestion
    state.Clear(alarms);
```
Unless I'm just missing something?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:63]: Please fix the restyle issues?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:1]: Some thought on the implementation:

1.  This seems to assume applications will directly read/write the attributes in the attribute store.  That's not ideal, because the fact that we are storing them in the attr store is a a cluster implementation detail.  We should have APIs for manipulating these attributes that the cluster exposes.
2. Should we be checking for unknown bits in the bitmaps?  The spec does not seem to say what should be done with unknown bits, which may need a spec issue.
3. When the attributes change we should emit the events the cluster requires to be emitted.  This will be easier to do if attribute changes go through APIs we expose.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:46]: Why not using the actual status?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:1]: > I was looking into how can we expose APIs for attribute read/write.

You can define functions in the cluster namespace in the cluster header, for example, that applications will call.

Or define an object that represents an instance of the cluster, and that has methods (and then you implement those methods in the cluster cpp file).

> If I'm not wrong, emissions of events will be part of driver here?

I'm not sure what you mean by "driver".

Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:1]: Added read/write APIs for attributes

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Updated

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:51]: If the get failed, isn't this printing garbage?  Same for the other helpers.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:85]: How is this supposed to work?   This code was never executed, clearly.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:90]: Why do you need to compare the raw values?   I'd expect operator== to work these objects.

Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:90]: Why are we doing this skip-the-set for mask but not the others?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:43]: Why all the `chip::` prefixes everywhere?  You have `using namespace chip;` above....

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:114]: You have `using namespace chip::app;` above too...

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:151]: Return false will also try to add status.  Never return false.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:142]: eActive already has everything cleared.  This logic does not make sense.  In fact, nothing ever becomes active _in this function_.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:143]: This is not making sense either.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Updated logic acc to spec

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Updated logic acc to spec

**Comment by marktrayer** [comment]
Changes needed to line up with approved (and merged) spec changes.  Please see:
https://github.com/project-chip/connectedhomeip/issues/26745

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Changes needed, per @marktrayer's comment.

**Comment by bzbarsky-apple** [comment]
@jadhavrohit924 Note https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/7107 in terms of what goes in the events.

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple This implementation is based latest spec changes. I have already added #27030 to fix xml. Here, CI will fail coz this branch don’t have updated zap generated code.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: I have added API to _set_ the state value

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: State in event is not updated in spec. It should be state after the change. I have updated it

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Added setter API for state

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: As per updated spec Reset command should not be part of refrigerator alarm cluster. I have removed this part.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:99]: This is wrong.  This should be "state except the things that were already active".

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:100]: And this should be "things that were active but are not in state".

Was this tested at all?  We should _really_ have some tests for this....

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:103]: This should be just `BitMask<AlarmMap> newState = state`, right?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:100]:
```suggestion
    BitMask<AlarmMap> becameInactive = ~state.Raw();
```
"inactive" is one word.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:139]: Per spec, an application is supposed to be able to decide whether it can support disabling the alarm, for the generic alarms clusters.

But specifically for refrigerator alarm, remote-disabling the alarm is disallowed, no?  I don't see that being enforced here.  Again: we need _tests_.  This part is trivial to test in a YAML.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h:44]: Argument names should match the C++ here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:152]: > But specifically for refrigerator alarm, remote-disabling the alarm is disallowed, no?

Spec says you cannot disarm ( alarm Active --> Inactive) the alarm which is triggered unless there is some physical action. It doesn’t mention you cannot enable/disable alarms?

I mean remote modification in `State` ( Active/Inactive ) is not allowed. But spec doesn’t mentioned about the `Mask` ( enable/disable ).

Comment (marktrayer) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:152]: The spec is pretty explicit that the mask is fixed, and disabling is only possible at the device itself.  Although as there does appear to be confusion, I'll make that clearer.

Comment (bzbarsky-apple) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:152]: @jadhavrohit924 Yes, what @marktrayer said.  The Mask value for this cluster is a fixed constant that cannot be changed.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: One interesting question:  If the Mask changes to suppress a now-active alarm, what should happen to the State value?  Should it change?  @marktrayer

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:110]: Why are you insisting on using bit-ops on Raw()?  How about:
```suggestion
    becameActive.Set(newState).Clear(currentState);
```

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:112]:
```suggestion
    becameInactive.Set(currentState).Clear(newState);
```

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:87]: How about we just call the second arg newState and then it will be clear that it's the new state and we don't have to have a separate `newState` variable below?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:90]: Might be better as oldState, but don't feel too strongly about it.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:130]: This line is wrong.  Was this tested at all?

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h:37]: This should document that it does SendNotifyEvent.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h:39]: Should this be public API?  Will someone ever need to do this without calling SetStateValue?

Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h]: Made it private

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: Are we suppose to change the `State` when `Mask` changes? Spec says `State` cannot to change unless there is any physical action at the device ( closing the door ). Isn’t it conflicting ?

> the server allows the alarm to be remotely suppressed, and the modify command would suppress that alarm, the alarm should be reset.

Remotely alarm cannot be suppressed for this cluster.

Comment (bzbarsky-apple) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: So when would Mask ever change for this cluster?  The answer is: only when physical action on the device suppresses the alarm, right?

And if physical action on the device suppresses the alarm, what should happen to State?

But more fundamentally: should we allow the cluster to get into a state where an alarm is:

1. Marked as suppressed in Mask.
2. Marked as currently alarming in State.

?  I would think we should not.  That means modifying State as needed when Mask changes, and perhaps rejecting State changes that don't match Mask or something...  @marktrayer @jadhavrohit924

Comment (marktrayer) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: So there's two things to keep clean here I think:
1. the base cluster (assuming we have that implemented as a fundamental anywhere) supports Mask modification, and if you modify a mask to suppress an active alarm, then that alarm is suppressed and the state shall change.
2.  refrigerator alarm specifically doesn't support Mask modification, but if the alarm is physically suppressed locally (via some magic per device sequence), then state changes.

Comment (bzbarsky-apple) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: > refrigerator alarm specifically doesn't support Mask modification

@marktrayer Including via physical interaction with the device?

Comment (marktrayer) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:72]: The note from @jadhavrohit924 is correct.  If the physical action suppresses the alarm, then the mask will change and the state will also change.  If subsequent physical action re-enables the alarm, the mask bit will be set.

Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Changed state when mask changes

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:86]: If this fails, do you really want to SetStateValue with potential garbage?

Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: This should not happen. Handlled the case.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp
Comment (bzbarsky-apple) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp:93]: So this is going to generate notify events and attribute reports for the state, etc, even if the state has not actually changed, right?  We should probably skip making this SetStateValue call if the mask change does not in fact affect the state....

Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.cpp]: Sure, done.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h
Comment (tcarmelveilleux) [inline src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h:30]: Why does this have to be singleton?

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter
Comment (tcarmelveilleux) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:2378]: There is a missing Latch attribute, which means the RESET feature is not implementable. Is it indended that the RESET feature cannot be implemented with this version of server cluster?

**Review by tcarmelveilleux — APPROVED** [review]

**Review by marktrayer — APPROVED** [review]

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.matter
Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:2378]: Spec says that RESET is always false in refrigerator alarm cluster.

File: src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h
Comment (jadhavrohit924) [reply src/app/clusters/refrigerator-alarm-server/refrigerator-alarm-server.h:30]: All the cluster implementation creates Server class as singleton class. I'm not sure about the reason, do we need to change it?

File: src/app/zap_cluster_list.json
Comment (jadhavrohit924) [reply src/app/zap_cluster_list.json:265]: It is SOMETHING_CLUSTER. Actually we need to fix cluster name in refrigerator-alarm.xml and regen zap so I'll add follow up PR on this

**Review by jmeg-sfy — APPROVED** [review]

**Comment by jmeg-sfy** [comment]
do we want to wait for the addition of the new support attribute ?
@jadhavrohit924

**Comment by jadhavrohit924** [comment]
@jmeg-sfy I don’t understand, what is new support attribute? Is there anything going on about refrigerator alarm cluster that I am not following?

**Comment by marktrayer** [comment]
@jadhavrohit924 there's spec churn happening.  There'll be a new issue raised to track that in the SDK once merged on the spec side.  I wouldn't hold this PR up waiting on that.

**Comment by jadhavrohit924** [comment]
Thanks @marktrayer. So if spec changes we have to change cluster definition in xml and regen zap too. I think we can have followup PRs when the spec changes gets merged.

**Pull request merged**
